gax/src/main/java/com/google/api/gax/retrying/BasicRetryingFuture.java

```
@@ -0,0 +1,216 @@
+  // getAttemptResult() call will return a new future, tracking the new attempt. Otherwise
+  // attemptResult is set to the same result as the one returned by peekAttemptResult(), indicating
+  // that the ultimate unmodifiable result of the whole future was reached.
+  private void setAttemptResult(
+      boolean cancelled, Throwable throwable, ResponseT response, boolean shouldRetry) {
+    ApiFuture<ResponseT> prevAttemptResult = attemptResult;
+    try {
+      if (cancelled) {
+        NonCancellableFuture<ResponseT> future = new NonCancellableFuture<>();
+        future.cancelPrivately();
+        latestCompletedAttemptResult = future;
+        attemptResult = shouldRetry ? null : latestCompletedAttemptResult;
+        if (prevAttemptResult instanceof NonCancellableFuture) {
+          ((NonCancellableFuture<ResponseT>) prevAttemptResult).cancelPrivately();
+        }
+      } else if (throwable != null) {
+        latestCompletedAttemptResult = ApiFutures.immediateFailedFuture(throwable);
+        attemptResult = shouldRetry ? null : latestCompletedAttemptResult;
+        if (prevAttemptResult instanceof NonCancellableFuture) {
+          ((NonCancellableFuture<ResponseT>) prevAttemptResult).setException(throwable);
+        }
+      } else {
+        latestCompletedAttemptResult = ApiFutures.immediateFuture(response);
+        attemptResult = shouldRetry ? null : latestCompletedAttemptResult;
+        if (prevAttemptResult instanceof NonCancellableFuture) {
+          ((NonCancellableFuture<ResponseT>) prevAttemptResult).set(response);
+        }
+      }
+    } catch (Throwable e) {
+      // Usually should not happen but is still possible, for example if one of the attempt callbacks
+      // throws an exception. An example of such condition is the OperationFuture
+      // which uses ApiFutures.transform(), which actually assign callbacks to the attempt result
+      // futures, and those can fail, for example if metadata class is a wrong one.
```

Add a note about why the exception is swallowed